src/mesh/interpolation/petsc_hermite_spline_xz.cxx

```diff
@@ -0,0 +1,240 @@
+    weights(i, i_c.xp(2).zm()) = -(h11_x[i] * h10_z[i] / 4);
+    weights(i, i_c.xp(2).zp()) = (h11_x[i] * h01_z[i] / 2) + (h11_x[i] * h10_z[i] / 4);
+    weights(i, i_c.xp(2).zp(2)) = (h11_x[i] * h11_z[i] / 4);
+    weights(i, i_c) = (h00_x[i] * h00_z[i]) + (h11_x[i] * h11_z[i] / 4)
+                      - (h00_x[i] * h11_z[i] / 2) - (h11_x[i] * h00_z[i] / 2);
+    weights(i, i_c.zm()) = (h11_x[i] * h10_z[i] / 4) - (h00_x[i] * h10_z[i] / 2);
+    weights(i, i_c.zp()) = (h00_x[i] * h01_z[i]) + (h00_x[i] * h10_z[i] / 2)
+                           - (h11_x[i] * h01_z[i] / 2) - (h11_x[i] * h10_z[i] / 4);
+    weights(i, i_c.zp(2)) = (h00_x[i] * h11_z[i] / 2) - (h11_x[i] * h11_z[i] / 4);
+  }
+
+  weights.assemble();
+}
+
+void PetscXZHermiteSpline::calcWeights(const Field3D& delta_x, const Field3D& delta_z,
+                                       const BoutMask& mask, const std::string& region) {
+  setMask(mask);
+  calcWeights(delta_x, delta_z, region);
+}
+
+/*!
+ * Return position and weight of points needed to approximate the function value at the
+ * point that the field line through (i,j,k) meets the (j+1)-plane. For the case where
+ * only the z-direction is not aligned to grid points, the approximation is: f(i,j+1,k*) =
+ * h00_z * f(i,j+1,k) + h01_z * f(i,j+1,k+1)
+ *                 + h10_z * dfdz(i,j+1,k) + h11_z * dfdz(i,j+1,k+1)
+ *               = h00_z * f(i,j+1,k) + h01_z * f(i,j+1,k+1)
+ *                 + h10_z * ( f(i,j+1,k+1) - f(i,j+1,k-1) ) / 2
+ *                 + h11_z * ( f(i,j+1,k+2) - f(i,j+1,k) ) / 2
+ * for k* a point between k and k+1. Therefore, this function returns
+ *   position 		weight
+ *   (i, j+1, k-1)	- h10_z / 2
+ *   (i, j+1, k)	h00_z - h11_z / 2
+ *   (i, j+1, k+1)	h01_z + h10_z / 2
+ *   (i, j+1, k+2)	h11_z / 2
+ */
+std::vector<ParallelTransform::PositionsAndWeights>
+PetscXZHermiteSpline::getWeightsForYApproximation(int i, int j, int k, int yoffset) {
+  const int nz = localmesh->LocalNz;
+  const int k_mod = k_corner(i, j, k);
+  const int k_mod_m1 = (k_mod > 0) ? (k_mod - 1) : (nz - 1);
+  const int k_mod_p1 = (k_mod == nz) ? 0 : k_mod + 1;
+  const int k_mod_p2 = (k_mod_p1 == nz) ? 0 : k_mod_p1 + 1;
+
+  return {{i, j + yoffset, k_mod_m1, -0.5 * h10_z(i, j, k)},
+          {i, j + yoffset, k_mod, h00_z(i, j, k) - 0.5 * h11_z(i, j, k)},
+          {i, j + yoffset, k_mod_p1, h01_z(i, j, k) + 0.5 * h10_z(i, j, k)},
+          {i, j + yoffset, k_mod_p2, 0.5 * h11_z(i, j, k)}};
+}
+
+Field3D
+PetscXZHermiteSpline::interpolate(const Field3D& f,
+                                  [[maybe_unused]] const std::string& region) const {
+
+  ASSERT1(f.getMesh() == localmesh);
+  const PetscVector<Field3D> f_petsc(f, indexer);
+  const PetscVector<Field3D> f_result = weights * f_petsc;
+  // TODO: we should only consider the passed-in region
+  // const auto region2 = y_offset == 0 ? region : fmt::format("RGN_YPAR_{:+d}", y_offset);
+  return f_result.toField();
```

Comment:
This does not set the region - is this set in fci.cxx/fci.hxx?
Otherwise this will not work with high check level and 3d metrics, afaik.